banyand/liaison/grpc/property.go

```diff
@@ -500,7 +533,152 @@ func (ps *propertyServer) remove(ids [][]byte) error {
 	return nil
 }
 
+func (ps *propertyServer) startRepairQueue(stop chan struct{}) {
+	ps.repairQueue = newRepairQueue(ps, ps.repairQueueCount)
+	ps.repairQueue.Start(stop)
+}
+
 type propertyWithMetadata struct {
 	*propertyv1.Property
 	deletedTime int64
 }
+
+type propertyWithCount struct {
+	*propertyWithMetadata
+	existNodes map[string]bool
+}
+
+func newPropertyWithCounts(p *propertyWithMetadata, existNode string) *propertyWithCount {
+	res := &propertyWithCount{
+		propertyWithMetadata: p,
+	}
+	res.addExistNode(existNode)
+	return res
+}
+
+func (p *propertyWithCount) addExistNode(node string) {
+	if p.existNodes == nil {
+		p.existNodes = make(map[string]bool)
+	}
+	p.existNodes[node] = true
+}
+
+type repairQueue struct {
+	queue         chan *repairTask
+	inProcess     map[repairInProcessKey]struct{}
+	ps            *propertyServer
+	processLocker sync.Mutex
+}
+
+func newRepairQueue(ps *propertyServer, size int) *repairQueue {
+	return &repairQueue{
+		queue:     make(chan *repairTask, size),
+		inProcess: make(map[repairInProcessKey]struct{}, size),
+		ps:        ps,
+	}
+}
+
+func (rq *repairQueue) addProperty(entity string, shardID uint, p *propertyWithCount) bool {
+	rq.processLocker.Lock()
+	defer rq.processLocker.Unlock()
+	key := repairInProcessKey{
+		entity:  entity,
+		modTime: p.Metadata.ModRevision,
+	}
+	if _, ok := rq.inProcess[key]; ok {
+		return false
+	}
+	rq.inProcess[key] = struct{}{}
+	task := &repairTask{
+		shardID:  shardID,
+		entity:   entity,
+		property: p,
+	}
+	// adding the task to the queue, if the queue is full, then ignore it
+	select {
+	case rq.queue <- task:
+	default:
+		return false
+	}
+	return true
+}
+
+func (rq *repairQueue) Start(stop chan struct{}) {
+	ctx, cancel := context.WithCancel(context.Background())
+	go func() {
+		<-stop
+		cancel()
+	}()
```

Comment:
No need to use this technology. The pipeline.publish has an internal timeout mechanism that is not affected by the context. You can simply use the stop channel to exit the goroutine of the task execution loop.